lib/Dialect/MSFT/MSFTPasses.cpp

```diff
@@ -247,6 +248,176 @@ std::unique_ptr<Pass> createLowerToHWPass() {
+    // Add the symbol to the cache.
+    cache.addDefinition(symName, &op);
+  }
+  cache.freeze();
+}
+
+void PartitionPass::runOnOperation() {
+  ModuleOp outerMod = getOperation();
+  ::populateSymbolCache(outerMod, topLevelSyms);
+
+  // TODO: sort modules by number of instantiations so we only have to run one
+  // "bubbling" pass.
+  outerMod.walk([this](MSFTModuleOp mod) { partition(mod); });
+}
+
+void PartitionPass::partition(MSFTModuleOp mod) {
+  DenseMap<SymbolRefAttr, SmallVector<Operation *, 4>> partMembers;
+  mod.walk([&partMembers](Operation *op) {
+    if (auto part = op->getAttrOfType<SymbolRefAttr>("targetDesignPartition"))
+      partMembers[part].push_back(op);
+  });
+
+  SmallVector<DesignPartitionOp, 2> parts(mod.getOps<DesignPartitionOp>());
+  for (auto part : parts) {
+    SymbolRefAttr partSym = SymbolRefAttr::get(SymbolTable::getSymbolName(mod),
+                                               {SymbolRefAttr::get(part)});
+
+    auto usersIter = partMembers.find(partSym);
+    if (usersIter != partMembers.end()) {
+      this->partition(part, usersIter->second);
+      partMembers.erase(usersIter);
+    }
+    part.erase();
+  }
+
+  // TODO: For operations which target partitions not in the same module, bubble
+  // them up.
+}
+
+void PartitionPass::partition(DesignPartitionOp partOp,
+                              SmallVectorImpl<Operation *> &toMove) {
+
+  auto *ctxt = partOp.getContext();
+  auto loc = partOp.getLoc();
+
+  //*************
+  //   Determine the partition module's interface. Keep some bookkeeping around.
+  SmallVector<hw::PortInfo, 128> ports;
+  SmallVector<Value, 64> partInstInputs;
+  SmallVector<Value, 64> partInstOutputs;
+
+  // Handle module-like operations. Not strictly necessary, but we can base the
+  // new portnames on the portnames of the instance being moved and the instance
+  // name.
+  auto addModuleLike = [&](Operation *inst, hw::ModulePortInfo modPorts) {
+    StringAttr name = SymbolTable::getSymbolName(inst);
+
+    for (auto port :
+         llvm::concat<hw::PortInfo>(modPorts.inputs, modPorts.outputs)) {
+      ports.push_back(
+          hw::PortInfo{/*name*/ StringAttr::get(ctxt, name.getValue() + "_" +
+                                                          port.name.getValue()),
+                       /*direction*/ port.direction,
+                       /*type*/ port.type,
+                       /*argNum*/ ports.size()});
+      if (port.direction == hw::PortDirection::OUTPUT)
+        partInstOutputs.push_back(inst->getResult(port.argNum));
+      else
+        partInstInputs.push_back(inst->getOperand(port.argNum));
+    }
+  };
+  // Handle all other operators.
+  auto addOther = [&](Operation *op) { assert(false && "Unimplemented"); };
+
+  // Aggregate the args/results into partition module ports.
+  for (Operation *op : toMove) {
+    if (auto inst = dyn_cast<InstanceOp>(op)) {
+      Operation *modOp = topLevelSyms.getDefinition(inst.moduleNameAttr());
```

Comment:
Ideally, `msft.instance` would know the port names (like `hw.instance`) but it's just not that advanced. If we ever get a `ModuleLike` interface, this would be great.